yang-check.sh

```diff
@@ -0,0 +1,177 @@
+  echo "    YANG-TREE <mod.yang> (pyang lint)"
+  echo "    YANG-DATA <mod.yang> <example.json> (yanglint validate json)"
+  echo "  uses rsync to download imported modules to modules/ from iana"
+  echo "error: no input file: '$input_md'"
+  exit 1
+fi
+
+modules=$(grep YANG-MODULE $input_md | awk '{print $2;}' | tr "\n" " ")
+tree_modules=$(grep YANG-TREE $input_md | awk '{print $2;}' | tr "\n" " ")
+data_modules_str=$(grep YANG-DATA $input_md | awk '{print $2;}' | tr "\n" " ")
+
+data_json_str="$(grep YANG-DATA $input_md | awk '{print $3;}' | tr "\n" " ")"
+all_modules="$(echo "$modules $tree_modules $data_modules_str" | tr " " "\n" | sort | uniq | tr "\n" " " | sed -e 's/^ *//' | sed -e 's/ *$//')"
+
+data_jsons=(); for d in $data_json_str; do data_jsons+=("$d"); done
+data_modules=(); for d in $data_modules_str; do data_modules+=("$d"); done
+
+#echo "input=$input_md, data='${#data_jsons[@]}: $data_jsons', mods='$all_modules'"
+
+had_error=0
+
+target_xml="$(echo ${input_md} | sed -e 's/.md$/.xml/')"
+for ((i=0; i < ${#data_jsons[@]}; i++)); do
+  if [ ! -f "${data_jsons[i]}" ]; then
+    echo "no example data file ${data_jsons[i]}"
+    had_error=1
+  fi
+done
+for mod in ${all_modules}; do
+  if [ ! -f ${mod} ]; then
+    echo "no module file ${mod}"
+    had_error=1
+  fi
+done
+
+if [ "${had_error}" != "0" ]; then
+  echo "error: some files not found, exiting"
+  exit 1
+fi
+
+fatal=0
+if ! which pyang 2>&1 > /dev/null ; then
+  echo "error in $0: pyang required; please install\n(pip install pyang)"
+  fatal=1
+fi
+
+if ! which yanglint 2>&1 > /dev/null ; then
+  echo "error in $0: yanglint required; please install libyang:"
+  echo "(install libyang-dev or equivalent if available, or build:)"
+  echo " git clone https://github.com/CESNET/libyang"
+  echo " mkdir libyang/build"
+  echo " pushd libyang/build"
+  echo " cmake .."
+  echo " make && sudo make install"
+  echo " popd"
+  echo " yanglint --help"
+  echo "if it can't load libyang.so, try something like:"
+  echo " cmake -DCMAKE_INSTALL_PREFIX=$$HOME/local-installs -DCMAKE_INSTALL_RPATH=$$HOME/local-installs/lib .."
+  fatal=1
+fi
+
+if [ "$fatal" != "0" ]; then
+  exit 1
+fi
+
+set -e
+
+if [ ! -d modules/ ]; then
+  mkdir -p modules/
+fi
+
+before=$(ls modules/ | wc -l)
+after=${before}
+last=$((before-1))
+
+echo "checking for module dependencies (downloading into modules/...)"
+while [ "${last}" != "${after}" ]; do
+  last=${after}
+  for mod in ${all_modules}; do
+    missing=$( ( pyang --verbose --path modules:. ${mod} 2>&1 || true ) | \
+      grep "error: module" | \
+      sed -e 's/.*: error: module "\([^"]*\)" not found in search path.*/\1/')
+    if [ "$missing" != "" ]; then
+      if ! which rsync 2>&1 > /dev/null ; then
+        echo "error in $0: rsync required; please install"
+      fi
+
+      for mis in ${missing}; do
+        bash -e -x -c "rsync -cvz rsync.iana.org::assignments/yang-parameters/${mis}*.yang modules/"
```

Comment:
I can use rsync directly, sure.  I was doing this mainly so that when rsync errors, the call that caused it will be visible, but the quoting is a very good point and I think it should echo instead, thanks.